drivers/interrupt_controller/intc_arcv2_irq_unit.c

```diff
@@ -21,43 +21,132 @@
+
+	return 0;
+}
+
+SYS_INIT(arc_shared_intc_update_post_smp, SMP, 0);
+#endif /* CONFIG_ARC_CONNECT */
+
+
+/* lowest IRQ priority */
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
+#define ARC_IRQ_DEFAULT_PRIORITY ((CONFIG_NUM_IRQ_PRIO_LEVELS - 1) | _ARC_V2_IRQ_PRIORITY_SECURE)
+#else
+#define ARC_IRQ_DEFAULT_PRIORITY (CONFIG_NUM_IRQ_PRIO_LEVELS - 1)
+#endif
+
+static inline void arc_core_intc_init_nolock(uint32_t irq, uint32_t state)
+{
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_PRIORITY, ARC_IRQ_DEFAULT_PRIORITY);
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_TRIGGER, _ARC_V2_INT_LEVEL);
+	z_arc_v2_aux_reg_write(_ARC_V2_IRQ_ENABLE, state);
+}
+
 /*
- * @brief Initialize the interrupt unit device driver
- *
- * Initializes the interrupt unit device driver and the device
- * itself.
+ * Initialize the core private interrupt controller.
  *
- * Interrupts are still locked at this point, so there is no need to protect
- * the window between a write to IRQ_SELECT and subsequent writes to the
- * selected IRQ's registers.
+ * This function must be called on each CPU in case of SMP system.
  *
- * @return 0 for success
+ * NOTE: core interrupts are still globally disabled at this point (STATUS32.IE = 0), so there is
+ * no need to protect the window between a write to IRQ_SELECT and subsequent writes to the
+ * selected IRQ's registers with locks.
  */
-static int arc_v2_irq_unit_init(const struct device *unused)
+void arc_core_private_intc_init(void)
 {
-	ARG_UNUSED(unused);
-	int irq; /* the interrupt index */
+	/*
+	 * Interrupts from 0 to 15 are exceptions and they are ignored by IRQ auxiliary registers.
+	 * We skip those interrupt lines while setting up core private interrupt controller.
+	 */
+	BUILD_ASSERT(CONFIG_GEN_IRQ_START_VECTOR == 16);
 
-	/* Interrupts from 0 to 15 are exceptions and they are ignored
-	 * by IRQ auxiliary registers. For that reason we skip those
-	 * values in this loop.
+	/*
+	 * System with IDU case (most likely multi-core system):
+	 *  - disable private IRQs: they will be enabled with irq_enable before usage
+	 *  - enable shared (IDU) IRQs: their enabling / disabling is controlled via IDU, so we
+	 *    always pass them via core private interrupt controller.
+	 * System without IDU case (single-core system):
+	 *  - disable all IRQs: they will be enabled with irq_enable before usage
 	 */
-	for (irq = 16; irq < CONFIG_NUM_IRQS; irq++) {
+#ifdef CONFIG_ARC_CONNECT
+	for (uint32_t irq = CONFIG_GEN_IRQ_START_VECTOR; irq < ARC_CONNECT_IDU_IRQ_START; irq++) {
+		arc_core_intc_init_nolock(irq, _ARC_V2_INT_DISABLE);
+	}
 
-		z_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
-#ifdef CONFIG_ARC_SECURE_FIRMWARE
-		z_arc_v2_aux_reg_write(_ARC_V2_IRQ_PRIORITY,
-			 (CONFIG_NUM_IRQ_PRIO_LEVELS-1) |
-			 _ARC_V2_IRQ_PRIORITY_SECURE); /* lowest priority */
+	for (uint32_t irq = ARC_CONNECT_IDU_IRQ_START; irq < CONFIG_NUM_IRQS; irq++) {
+		arc_core_intc_init_nolock(irq, _ARC_V2_INT_ENABLE);
```

Comment:
Could there be any common interrupts pending at this time, that first need to get cleared? Maybe from a warm reset?

Reply:
Shouldn't be an issue for warm reset as we firstly disable common lines in `arc_shared_intc_init` and only after that we enable it here, in `arc_core_private_intc_init`.

Also shouldn't be an issue for hard reset - as after reset all lines in core interrupt controller are enabled anyway.